File: 20.py
# Runtime: 36 ms, faster than 98.90% of Python3 online submissions for Valid Parentheses.
# Difficulty: Easy

class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        opens = {'(':0, '{':1, '[':2}
        closes = {')':0, '}':1, ']':2}
        stack = list()
        
        for char in s:
            if char in opens:
                stack.append(opens[char])
            elif char in closes and not stack or closes[char] != stack.pop():
                return False
        return len(stack) == 0

File: 20.java
// Runtime: 2 ms, faster than 61.00% of Java online submissions for Valid Parentheses.
// Memory Usage: 34.2 MB, less than 100.00% of Java online submissions for Valid Parentheses.

import java.util.HashMap;

class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> openCloseMap = new HashMap<>();
        openCloseMap.put('(', ')');
        openCloseMap.put('{', '}');
        openCloseMap.put('[', ']');
        
        Stack<Character> parenthesis = new Stack<>();
        for (Character c : s.toCharArray()) {
            if (openCloseMap.containsKey(c)) {
                parenthesis.push(openCloseMap.get(c));
            } else {
                if (parenthesis.size() == 0 || !parenthesis.pop().equals(c))
                    return false;
            }
        }
        
        return parenthesis.size() == 0;
    }
}
